service/controller/clusterapi/v17/resources/tenantclients/create.go

```diff
@@ -3,67 +3,14 @@ package tenantclients
 import (
 	"context"
 
-	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
 	"github.com/giantswarm/microerror"
-	corev1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/client-go/kubernetes"
-	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
-
-	"github.com/giantswarm/cluster-operator/service/controller/clusterapi/v17/controllercontext"
-	"github.com/giantswarm/cluster-operator/service/controller/clusterapi/v17/key"
 )
 
 func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
-	var cluster v1alpha1.Cluster
-	var err error
-
-	// This resource is used both from Cluster and MachineDeployment
-	// controllers so it must work with both types.
-	switch obj.(type) {
-	case v1alpha1.Cluster:
-		cluster, err = key.ToCluster(obj)
-		if err != nil {
-			return microerror.Mask(err)
-		}
-
-	case v1alpha1.MachineDeployment:
-		cr, err := key.ToMachineDeployment(obj)
-		if err != nil {
-			return microerror.Mask(err)
-		}
-
-		m, err := r.cmaClient.ClusterV1alpha1().Clusters(corev1.NamespaceAll).Get(key.ClusterID(&cr), metav1.GetOptions{})
-		if err != nil {
-			return microerror.Mask(err)
-		}
-
-		cluster = *m
-	}
-
-	cc, err := controllercontext.FromContext(ctx)
+	err := r.ensure(ctx, obj)
```

Comment:
To improve the situation like we did in the `machinedepoymentstatus` resource I opted for the `ensure` method here as well. Both `EnsureCreated` and `EnsureDeleted` make use of it.